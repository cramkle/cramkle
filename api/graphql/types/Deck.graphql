# import Note from "Note.graphql"
# import User from "User.graphql"

type StudySessionDetails {
  newCount: Int!
  learningCount: Int!
  reviewCount: Int!
}

"""
Collection of notes
"""
type Deck implements Node {
  """
  Deck id
  """
  id: ID!
  """
  Title of the deck
  """
  title: String!
  """
  Description of the deck
  """
  description: String
  """
  Unique identifiable slug
  """
  slug: String!
  """
  Notes contained in this deck
  """
  notes: [Note]
  """
  Owner of the deck
  """
  owner: User
  """
  Whether this deck is published to the marketplace
  """
  published: Boolean
  """
  Details of current study session
  """
  studySessionDetails: StudySessionDetails!
}
