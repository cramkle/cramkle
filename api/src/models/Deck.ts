import { Document, Schema, model } from 'mongoose'
import shortid from 'shortid'

interface Deck {
  title: string
  description?: string
  slug: string
  published: boolean
}

interface DeckDocument extends Deck, Document {}

const DeckSchema = new Schema<DeckDocument>({
  title: {
    type: String,
    required: true,
  },
  description: String,
  slug: {
    type: String,
    unique: true,
  },
  notes: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Note',
    },
  ],
  ownerId: {
    type: Schema.Types.ObjectId,
    ref: 'User',
  },
  published: Boolean,
})

DeckSchema.pre('save', function (next) {
  const deck = this as DeckDocument

  if (!deck.isNew) {
    return next()
  }

  deck.slug = shortid.generate()

  next()
})

export default model<DeckDocument>('Deck', DeckSchema)
