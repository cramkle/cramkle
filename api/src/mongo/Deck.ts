import { Document, Schema, Types, model } from 'mongoose'
import shortid from 'shortid'

export interface Deck {
  title: string
  description?: string
  slug: string
  published: boolean
  ownerId: Types.ObjectId
  notes: Types.Array<Types.ObjectId>
}

export interface DeckDocument extends Deck, Document {}

const DeckSchema = new Schema<DeckDocument>({
  title: {
    type: String,
    required: true,
  },
  description: String,
  slug: {
    type: String,
    unique: true,
  },
  notes: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Note',
    },
  ],
  ownerId: {
    type: Schema.Types.ObjectId,
    ref: 'User',
  },
  published: Boolean,
})

DeckSchema.pre('save', function (next) {
  const deck = this as DeckDocument

  if (!deck.isNew) {
    return next()
  }

  deck.slug = shortid.generate()

  next()
})

export default model<DeckDocument>('Deck', DeckSchema)
