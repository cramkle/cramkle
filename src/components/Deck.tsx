import React from 'react'
import { withRouter, RouteComponentProps } from 'react-router'
import Card, { CardPrimaryContent } from '@material/react-card'
import { Body2, Headline6 } from '@material/react-typography'

interface Props extends RouteComponentProps {
  title: string
  description?: string
  slug: string
}

const Deck: React.FunctionComponent<Props> = ({ title, description, slug, history }) => (
  <Card outlined className="h-100">
    <CardPrimaryContent
      className="pa2 h-100"
      onClick={() => history.push(`/d/${slug}`)}
    >
      <Headline6>{title}</Headline6>
      {description && <Body2>{description}</Body2>}
    </CardPrimaryContent>
  </Card>
)

Deck.defaultProps = {
  description: null,
}

export default withRouter(Deck)
