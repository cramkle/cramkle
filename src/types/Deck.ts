export interface IDeck {
  id: string
  slug: string
  title: string
  description?: string
}
