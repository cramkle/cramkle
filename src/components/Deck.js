import * as React from 'react'
import Card, { CardPrimaryContent } from '@material/react-card'

import { DeckEntity } from '../types'

interface DeckProps extends DeckEntity {}

const Deck: React.StatelessComponent<DeckProps> = ({ title, description }) => (
  <Card outlined>
    <CardPrimaryContent style={{ padding: '.3rem .5rem' }}>
      <h3 className="mdc-typography--headline6">{title}</h3>
      <p className="mdc-typography--body2">{description}</p>
    </CardPrimaryContent>
  </Card>
)

Deck.defaultProps = {
  description: null,
}

export default Deck
