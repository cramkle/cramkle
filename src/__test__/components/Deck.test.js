import React from 'react'
import { render } from 'react-testing-library'

import Deck from '../../components/Deck'

describe('<Deck />', () => {
  it('should match snapshot', () => {
    const comp = render(<Deck title="日本語げんき" />)

    expect(comp.container).toMatchSnapshot()
  })
})
