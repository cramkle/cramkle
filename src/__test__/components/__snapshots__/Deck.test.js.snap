// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Deck /> should match snapshot 1`] = `
<div>
  <div
    class="mdc-card mdc-card--outlined"
  >
    <div
      class="mdc-card__primary-action"
      style="padding: .3rem .5rem;"
    >
      <h3
        class="mdc-typography--headline6"
      >
        日本語げんき
      </h3>
      <p
        class="mdc-typography--body2"
      />
    </div>
  </div>
</div>
`;
